### PR TITLE
[mlir][MemRef] Add ExtractStridedMetadataOpCollapseShapeFolder

### PR DESCRIPTION
**Pull request by dcaballe**
This PR adds a new pattern to the set of patterns used to resolve the offset, sizes and stride of a memref. Similar to `ExtractStridedMetadataOpSubviewFolder`, the new pattern resolves strided_metadata(collapse_shape) directly, without introduce a reshape_cast op.

### CODE REVIEW

**Comment by dcaballe** [comment]
Interesting enough, I can't find where `memref::populateExpandStridedMetadataPatterns` is being tested???

**Comment by qcolombet** [comment]
> Interesting enough, I can't find where `memref::populateExpandStridedMetadataPatterns` is being tested???

`mlir/test/Dialect/MemRef/expand-strided-metadata.mlir`, no?

File: mlir/lib/Dialect/MemRef/Transforms/ExpandStridedMetadata.cpp [hunk 1]
```diff
@@ -550,6 +550,78 @@ getCollapsedStride(memref::CollapseShapeOp collapseShape, OpBuilder &builder,
   return {makeComposedFoldedAffineMin(builder, collapseShape.getLoc(), minMap,
                                       groupStrides)};
 }
+
+template <typename ReassociativeReshapeLikeOp,
```
Comment (qcolombet) [inline mlir/lib/Dialect/MemRef/Transforms/ExpandStridedMetadata.cpp:554]: Could you move the comment explaining what this does?

**Comment by qcolombet** [comment]
> > Interesting enough, I can't find where `memref::populateExpandStridedMetadataPatterns` is being tested???
> 
> `mlir/test/Dialect/MemRef/expand-strided-metadata.mlir`, no?

I guess you meant `populateResolveExtractStridedMetadataPatterns` since that's the one you're modifying.
I thought that `populateResolveExtractStridedMetadataPatterns` was a subset of `populateExpandStridedMetadataPatterns` but looks like it changed with https://reviews.llvm.org/D147393 and I didn't notice.

I see two paths forward
1. use `populateResolveExtractStridedMetadataPatterns` inside `populateExpandStridedMetadataPatterns` so that we can be sure `Resolve` is a subset of `Expand`. I think this is true today (modulo what https://reviews.llvm.org/D147393 added, but re-converging remains good IMHO) and we can revisit when they need to diverge.
2. integrate a test in the narrowing type test pass, since this is the only place where `Resolve` is used.

**Review by qcolombet — APPROVED** [review]
LGTM modulo:
1. add a test
2. move some comments on the refactored helper function

File: mlir/lib/Dialect/MemRef/Transforms/ExpandStridedMetadata.cpp [hunk 2]
```diff
@@ -550,6 +550,78 @@ getCollapsedStride(memref::CollapseShapeOp collapseShape, OpBuilder &builder,
...
+          SmallVector<OpFoldResult> (*getReshapedSizes)(
+              ReassociativeReshapeLikeOp, OpBuilder &,
+              ArrayRef<OpFoldResult> /*origSizes*/, unsigned /*groupId*/),
+          SmallVector<OpFoldResult> (*getReshapedStrides)(
+              ReassociativeReshapeLikeOp, OpBuilder &,
+              ArrayRef<OpFoldResult> /*origSizes*/,
+              ArrayRef<OpFoldResult> /*origStrides*/, unsigned /*groupId*/)>
```
Comment (ftynse) [inline mlir/lib/Dialect/MemRef/Transforms/ExpandStridedMetadata.cpp:561]: A fly-by nit: is there an observable benefit to using these template arguments as opposed to passing in `function_ref` callbacks?

**Comment by dcaballe** [comment]
> I see two paths forward
> 
> 1. use `populateResolveExtractStridedMetadataPatterns` inside `populateExpandStridedMetadataPatterns` so that we can be sure `Resolve` is a subset of `Expand`. I think this is true today (modulo what [reviews.llvm.org/D147393](https://reviews.llvm.org/D147393) added, but re-converging remains good IMHO) and we can revisit when they need to diverge.
> 2. integrate a test in the narrowing type test pass, since this is the only place where `Resolve` is used.

Thanks for the suggesting. Ok, let me try 1.

File: mlir/lib/Dialect/MemRef/Transforms/ExpandStridedMetadata.cpp (hunk 2 above)
Comment (qcolombet) [reply mlir/lib/Dialect/MemRef/Transforms/ExpandStridedMetadata.cpp:561]: I could be misremembering, but I think function_ref doesn't work for template arguments. (Aka the compiler chokes on them.)

**Pull request merged**